### DISCUSSION TITLE
Network API

### DISCUSSION THREAD

**Discussion by grixu** [archisquad/repository#35] [RFCs]
## 1. Purpose

The purpose of this RFC is to lay the foundation for the next API, which constitutes the entirety of the functioning repository library.

## 2. Reasons for Change

In its essence, it is intended to serve as a separation point between the infrastructure layer and the domain layer. Since a significant portion of the logic related to the infrastructure layer of the PWA application is related to handling network requests, defining an API in our solution that focuses on controlling network requests based on the solution provided by the user is necessary.

## 3. Proposed Implementation
In general, network communication is a broad topic, and for standard applications written in JavaScript, there is an immense range of different ways of communication, protocols, and data serialization methods.

However, our PWA-focused library will support any library and communication method chosen by the user, along with the two most popular data serialization methods.

Performing network requests can be reduced to four basic actions: create, update, retrieve, and delete. In APIs based on the REST pattern, these actions are differentiated based on the HTTP method used to make the request. However, in other solutions, such differentiation may be done in a completely different way, such as APIs communicating through WebSockets.

Nevertheless, the basic assumption regarding the four operations remains valid regardless of the method and communication protocol used. Therefore, to provide the best coverage for different cases users may encounter, the Network API would require the user to provide a configuration object containing four functions that execute these network requests.

```ts
interface CommunicationConfig<TOutCreate, TOutUpdate, TOutDelete> {
  get: () => Promise<unknown>
  create: (data: TOutCreate) => Promise<unknown>
  update: (data: TOutUpdate) => Promise<unknown>
  delete: (data: TOutDelete) => Promise<unknown>
}
```

The data received from the API is of type unknown because we cannot be certain of its type. Therefore, validation is often necessary, both at the TypeScript level and at runtime, to ensure type correctness.

In such cases, we need to introduce the ability to define functions that validate data and allow the passing of result types.

```ts
interface ValidationConfig<TGet, TCreate, TUpdate, TDelete> {
  get: (data: unknown) => Promise<TGet>
  create: (data: unknown) => Promise<TCreate>
  update: (data: unknown) => Promise<TUpdate>
  delete: (data: unknown) => Promise<TDelete>
}
```

Of course, data received from the API may have a completely different structure than that defined using the Entity API. This is where the separation of layers, or the use of hexagonal architecture in PWA applications to achieve separation between data models used in API services and data models used in the domain layer, comes into play. Mapping is applied to handle data transformation.

Mapping is generally a function that transforms one data format into another. It is challenging to define any automation that satisfies 100% of cases. Therefore, users should define mappers themselves.

```ts
interface MapperConfig<TEntity, TGet, TCreate, TUpdate, TDelete> {
  get: (data: TGet) => Promise<TEntity>
  create: (data: TEntity) => Promise<TCreate>
  update: (data: TEntity) => Promise<TUpdate>
  delete: (data: TEntity) => Promise<TDelete>
}
```

As seen in the mapping case, there is a certain directionality.

Of course, there is also room for deserialization in the entire process, which occurs after data retrieval. However, due to the lack of information about the method or library the user will use inside communication functions, we must allow for the existence of an empty serializer/deserializer.

Therefore, the entire configuration object would look like this:

```ts
interface NetworkApiConfig<TEntity, TGet, TCreate, TUpdate, TDelete, TOutCreate, TOutUpdate, TOutDelete> {
  communication: CommunicationConfig<TOutCreate, TOutUpdate, TOutDelete>
  validation: ValidationConfig<TGet, TCreate, TUpdate, TDelete>
  mappers: MapperConfig<TEntity, TGet, TCreate, TUpdate, TDelete>
  serializers: Serializers
}
```

### Batch Actions

The four basic actions described are sufficient, but some backends may be prepared to handle batch processing. Such endpoints will always be more efficient than creating numerous individual queries.

Furthermore, in the case of data retrieval, I suspect that 100% of APIs have an endpoint for batch retrieval of multiple entities.

Therefore, the CommunicationConfig interface should accept four optional additional functions handling batch processing endpoints:

```ts
interface CommunicationConfig<TOutCreate, TOutUpdate, TOutDelete> {
  get: () => Promise<unknown>
  create: (data: TOutCreate) => Promise<unknown>
  update: (data: TOutUpdate) => Promise<unknown>
  delete: (data: TOutDelete) => Promise<unknown>
  getMany?: () => Promise<unknown[]>
  createMany?: (data: TOutCreate[]) => Promise<unknown[]>
  updateMany?: (data: TOutUpdate[]) => Promise<unknown[]>
  deleteMany?: (data: TOutDelete[]) => Promise<unknown[]>
}
```

### Pagination

Another challenge is pagination, available in practically most APIs used for batch data retrieval. To retrieve the entire content from the API, we need to iterate through this process. This requires information about the page size and pagination method - two commonly used methods are page-based pagination and cursor-based pagination.

To enable pagination support, we can change the way of defining functions handling endpoints to declare them using a function that receives a context containing necessary information to handle this functionality:

```ts
interface CommunicationConfig {
  get: (context: CommunicationFnContext) => () => Promise<unknown>
  create: (context: CommunicationFnContext) => (data: unknown) => Promise<unknown>
  update: (context: CommunicationFnContext) => (data: unknown) => Promise<unknown>
  delete: (context: CommunicationFnContext) => (data: unknown) => Promise<unknown>
  getMany?: (context: CommunicationFnContext) => () => Promise<unknown>
  createMany?: (context: CommunicationFnContext) => (data: unknown) => Promise<unknown>
  updateMany?: (context: CommunicationFnContext) => (data: unknown) => Promise<unknown>
  deleteMany?: (context: CommunicationFnContext) => (data: unknown) => Promise<unknown>
}

interface CommunicationFnContext {
  currentPage?: number
  pageSize?: number
  cursor?: string
}
```

The advantage of this solution is that the context can be extended in the future by adding auxiliary functions or values.

### Background Data Fetching

The responsibility for this functionality lies with the user's chosen library for making API requests. Apart from that, usage strategies can be very different. By default, every request is made in the background - only using the await keyword suspends the execution of subsequent synchronous tasks in the queue until the Promise resolves.

However, for executing operations through the Network API, information about whether a particular request is executed in the background can be useful. Therefore, in the context of the function defining the API request, we will pass a function setting this value.

```ts
interface CommunicationFnContext {
  currentPage?: number
  pageSize?: number
  cursor?: string
  setRequestInBackground: () => void
}
```

### Data Returned from the Backend

Data returned from the backend may contain different sets of information than those sent to it. A simple example is assigning an identifier to a database entry. If the user needs this entry for some reason, it indicates an additional update of the Entity after resolving this network request.

**This issue can be addressed in two ways:**

A. Build such a mechanism in the Network API and give the user the choice of whether to update the Entity after executing create-update queries.

B. Create a universal hooks mechanism that provides the ability to perform any number of actions after the completion of each of the eight available network requests.

### Simplification of Mapping

As mentioned, mapping is a complex process, but in some cases, the API can be straightforward. For example, a BFF using the Entity data schema. In this case, we can skip mapping. Another example is using the Entity data schema, but the difference lies in casing. Backend fields use kebab_case, and Entity uses camelCase.

In such cases, the library could provide simple mappers that perform these simple mappings and save the user time and trouble.

### Custom Serializers
The two basic serializers available in the Network API system are JSON and Protocol Buffers. However, in both cases, the user may want to define additional actions that can occur before or after serialization/deserialization. Therefore, it is good to modify the accepted type in the configuration:

```ts
type Serializers = 'json' | 'protobuf' | 'none'
type CustomSerializer = (data: Buffer) => unknown

interface NetworkApiConfig<TEntity, TGet, TCreate, TUpdate, TDelete, TOutCreate, TOutUpdate, TOutDelete> {
  communication: CommunicationConfig<TOutCreate, TOutUpdate, TOutDelete>
  validation: ValidationConfig<TGet, TCreate, TUpdate, TDelete>
  mappers: MapperConfig<TEntity, TGet, TCreate, TUpdate, TDelete>
  serializers: Serializers | CustomSerializer
}
```

Allowing the user to define their own function provides a very flexible solution.

## 4. Drawbacks
N/A

## 5. Dependencies
No additional dependencies are introduced for libraries. Minimal dependencies are also created for the Entity API.

### Failed Requests

Handling failed requests also lies with the user and their chosen communication solution. The biggest obstacle to any intervention on our part is the multitude of possibilities. PWA, by design, uses Service Workers, but whether the developer decides to handle failed requests on this layer or in the main thread is up to them. Are unsuccessful requests automatically retried, or is there a time interval? Is there one retry strategy, or many? What are they dependent on? HTTP error code? And how is the offline state handled?

However, what we should implement is marking such a situation for the entity it concerns. There are at least two ways to do this:

**A. Embed another state into SyncMap**

```ts
export type SyncStatus = "not-synced" | "syncing" | "synced" | "failed"
```

**B. Build a separate tracking mechanism**

A separate mechanism would work similarly to the observer of effects and dependencies in the Vue 3 reactivity system. Built using WeakMap, there is no risk of memory leaks. The advantage is also the obvious separation of concerns. In the case of the SyncMap solution, this class will start to respond to more than one question - not only whether I have a fresh Entity instance but also which Entity did not synchronize. These are two different reasons for change, violating the Single Responsibility Principle. This topic requires a separate RFC.

### Soft Delete

The soft-delete mechanism is recommended to be implemented on the API responsible for storing entities. The repository, when selecting data, uses this API to determine the entities held in memory. Therefore, the handling of this mechanism should be there.

## 7. Summary
In summary, the Network API primarily consists of an orchestrator class that performs necessary network queries and accepts configuration within interfaces.

Additionally, it includes functions for serialization/deserialization to JSON and Protocol Buffers. And simple mappers that are useful in the case of simple differences in name casing between the API and entity data models.

Of course, the final form of the class or its eventual division is an implementation decision. However, the possibility and scope of coverage with functionalities are clearly expressed in this document. Unresolved issues will be addressed in subsequent RFCs.

**Comment by Mat3uszkek** [comment]
Great RFC! I just have one question about point 5 - Dependencies. Maybe it might be useful to specify what the minimal dependencies are?

**Reply by grixu** [reply]
Thanks for your input, @Mat3uszkek  👍 A dependency in this context means more like a dependency on the class level. I mentioned that no new libraries were needed for that part, but generally, we want to elaborate on those between the APIs we are creating. For now, Network API uses Entity API as an object - so IMHO, there isn't dependency between them at all. Because the code of Entity API has already been refactored to the version without SyncMap, we can use any object as input from a Network API perspective.

**Reply by grixu** [reply]
In terms of minimal dependencies required to run the repository library - it's the role of `package.json` to define that. So, as long there is nothing new here - I don't see any reason to mention it here. WDYT?
